### PR TITLE
Automatically set array size in  Stream benchmark

### LINKED ISSUE #120: Revisit the stream microbenchmark array sizes
The array sizes of the stream microbenchmarks should be changed to reflect the cache sizes of the new processors

### PR DESCRIPTION
**Pull request by teojgo**
Closes #120

### CODE REVIEW

File: checks/microbenchmarks/cpu/stream/stream.py [hunk 1]
```diff
@@ -67,16 +46,25 @@ def triad(self):
     @run_after('setup')
     def prepare_test(self):
         self.skip_if_no_procinfo()
+
+        # Sort the caches by type alphabetically (L1 < L2 < L3 ...) and get
+        # the total cache size of the last-level cache
+        caches = self.current_partition.processor.topology['caches']
+        last_level_cache = max(caches, key=lambda c: c['type'])
+        cache_size_bytes = (int(last_level_cache['size']) *
```
Comment (jgphpc) [inline checks/microbenchmarks/cpu/stream/stream.py:54]:
```suggestion
        cache_size_bytes = (int(last_level_cache['size']) *
```

File: checks/microbenchmarks/cpu/stream/stream.py [hunk 2]
```diff
@@ -67,16 +46,25 @@ def triad(self):
...
         self.skip_if_no_procinfo()
+
+        # Sort the caches by type alphabetically (L1 < L2 < L3 ...) and get
+        # the total cache size of the last-level cache
+        caches = self.current_partition.processor.topology['caches']
+        last_level_cache = max(caches, key=lambda c: c['type'])
+        cache_size_bytes = (int(last_level_cache['size']) * 
+                            len(last_level_cache['cpusets']))
+
+        # Sizes of each array must be at least 4x the size of the sum of all
```
Comment (jgphpc) [inline checks/microbenchmarks/cpu/stream/stream.py:57]:
```suggestion
        # Sizes of each array must be at least 4x the size of the sum of all
```

File: checks/microbenchmarks/cpu/stream/stream.py [hunk 3]
```diff
@@ -67,16 +46,25 @@ def triad(self):
     @run_after('setup')
     def prepare_test(self):
         self.skip_if_no_procinfo()
+
+        # Sort the caches by type alphabetically (L1 < L2 < L3 ...) and get
+        # the total cache size of the last-level cache
```
Comment (jgphpc) [inline checks/microbenchmarks/cpu/stream/stream.py:51]:
```suggestion
        # the total cache size of the last-level cache
        # for example: 
        # last_level_cache = {'type': 'L3', 'size': 33554432}
```

File: checks/microbenchmarks/cpu/stream/stream.py [hunk 4]
```diff
@@ -67,16 +46,25 @@ def triad(self):
...
+        array_size = 4 * cache_size_bytes // 8
+        self.build_system.cppflags = [f'-DSTREAM_ARRAY_SIZE={array_size}']
+
         self.num_cpus_per_task = int(self.current_partition.processor.num_cores)
         self.env_vars['OMP_NUM_THREADS'] = self.num_cpus_per_task
-
         self.build_system.cflags += (
             self.current_environ.extras.get('c_openmp_flags', ['-fopenmp'])
         )
         self.build_system.cflags.append('-O3')
```
Comment (jgphpc) [inline checks/microbenchmarks/cpu/stream/stream.py:67]: move all build_system up to line `self.build_system.cppflags`

File: checks/microbenchmarks/cpu/stream/stream.py [hunk 5]
```diff
@@ -67,16 +46,26 @@ def triad(self):
...
-        envname = self.current_environ.name
 
-        try:
-            self.reference = self.stream_bw_reference[envname]
-        except KeyError:
-            self.reference = self.stream_bw_reference['PrgEnv-gnu']
+        # This is typically needed for large array sizes
```
Comment (jgphpc) [inline checks/microbenchmarks/cpu/stream/stream.py:70]: should we keep the perf. reference ? this was working

**Review by ekouts — APPROVED** [review]

**Pull request merged**